kolena/_utils/instrumentation.py

```diff
@@ -87,3 +88,41 @@ def report_crash(id: int, endpoint_path: str):
     request = CoreAPI.MarkCrashedRequest(test_run_id=id)
     # note no krequests.raise_for_status -- already in crashed state
     krequests.post(endpoint_path=endpoint_path, data=json.dumps(dataclasses.asdict(request)))
+
+
+def set_profile():
+    krequests.put(endpoint_path=EventTrackingAPI.Path.PROFILE)
+
+
+def track_event(request: EventTrackingAPI.TrackEventRequest):
+    krequests.post(endpoint_path=EventTrackingAPI.Path.EVENT, data=json.dumps(dataclasses.asdict(request)))
+
+
+def with_invocation_tracked(event_name: str):
+    """function decorator to track start and end of an event"""
+
+    def tracking_decorator(func: Callable) -> Callable:
+        @functools.wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            if not _client_state.telemetry:
+                return func(*args, **kwargs)
+            # track start of event
+            track_event(EventTrackingAPI.TrackEventRequest(event_name=event_name + "-started"))
+
+            # track success or failure of the call, and if failed record the exception class name
+            try:
+                response = func(*args, **kwargs)
+                track_event(EventTrackingAPI.TrackEventRequest(event_name=event_name + "-succeeded"))
+                return response
+            except Exception as e:
+                track_event(
+                    EventTrackingAPI.TrackEventRequest(
+                        event_name=event_name + "-failed",
```

Comment:
What is the additional benefit of having started _and_ succeeded/failed events vs. just succeeded/failed events?

Reply:
I was following [similar strategy as marina](https://github.com/kolenaIO/marina/blob/87965f63965f20f75f870bdfc27255eedda692ad/src/lib/hooks/tracing/useTracking.ts#L203) which tracks both the start and end of the event. My hope was that we can track the duration of these activities. I think you are right though given that:
- This is inherently different from tracking front-end events where timing is controlled by the user. We are essentially just tracking API latency here.
- The mixpanel python client does not have a timing functionality similar to the [JS one](https://docs.mixpanel.com/docs/tracking/reference/javascript-full-api-reference#mixpaneltime_event), so we will need to handle the timing on our end.

Now I am leaning toward just tracking the event (without started/failed/succeeded). If needed, we can add metadata in the future to these events to include things like duration, exception message down the road. What do you think?

Reply:
That makes sense to me. I think keeping the events somewhat generic and just capturing additional metadata through custom event properties is the better approach 👍